File: Algo/Medium/0003_Longest_Substring_Without_Repeating_Characters/Solution.cpp
// 2022-06-10
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int longest = 0;
        bitset<256> seen;
        
        int left = 0, right = 0;
        while (right < s.length()) {
            while (seen[s[right]]) {
                seen[s[left]] = 0;
                left++;
            }
            seen[s[right]] = 1;
            right++;
            longest = longest > right-left ? longest : right-left;
        }
        
        return longest;
    }
};

File: Algo/Medium/0003_Longest_Substring_Without_Repeating_Characters/Solution.py
## 2022-08-23 (sliding window)
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        longest = 0
        length = len(s)
        
        contains = set()
        left, right = 0, 0
        while right < length:
            ## add values to window until we've run into a repeat
            if s[right] not in contains:
                contains.add(s[right])
                right += 1
            
            longest = max(longest, right - left)
            
            ## remove values from window until we've removed the repeat
            if right < length and s[right] in contains:
                contains.remove(s[left])
                left += 1
        
        return longest